projects/calc/js/main.js

```diff
@@ -0,0 +1,365 @@
+      if(!this._hasError || (this._hasError && e.keyCode === 27)) {
+        this._handleKeyboardSpecialClick(e);
+      }
+
+      this._showResult(this._result || 0);
+    }.bind(this));
+
+    // Show result by default
+    this._showResult(0);
+  }
+
+  // Handle button's click
+  _handleButtonsClick(e) {
+    let dataset = e.target.dataset;
+
+    // Button's handler
+    switch(dataset.type) {
+      case 'operand':
+        this._handleOperand(dataset.value);
+        break;
+      case 'operator':
+        this._handleOperator(dataset.value);
+        break;
+      case 'point':
+        this._handlePoint(dataset.value);
+        break;
+      case 'remove':
+        this._handleRemoveButton();
+        break;
+      case 'sign':
+        this._handleChangeSign();
+        break;
+      case 'result':
+        this._handleResult();
+        break;
+      case 'clear':
+        this._handleClear();
+        break;
+    }
+  }
+
+  // Handle keyboard's click
+  _handleKeyboardClick(e) {
+    let keyCode = e.keyCode,
+        char    = getChar(e);
+
+    // operand
+    if(keyCode >= 48 && keyCode <= 57) {
+      this._handleOperand(char);
+    }
+
+    // operator
+    if(keyCode === 42 || keyCode === 43 || keyCode === 45 || keyCode === 47) {
+      this._handleOperator(char);
+    }
+
+    // point
+    if(keyCode === 46) {
+      this._handlePoint(char);
+    }
+
+    function getChar(event) {
+      if(event.which == null) {
+        if(event.keyCode < 32) return null;
+        return String.fromCharCode(event.keyCode)
+      }
+
+      if(event.which != 0 && event.charCode != 0) {
+        if(event.which < 32) return null;
+        return String.fromCharCode(event.which);
+      }
+
+      return null;
+    }
+  }
+
+  _handleKeyboardSpecialClick(e) {
+    let keyCode = e.keyCode;
+
+    // clear
+    if(keyCode === 8) {
+      this._handleRemoveButton();
+    }
+
+    // result
+    if(keyCode === 13) {
+      this._handleResult();
+    }
+
+    // clear
+    if(keyCode === 27) {
+      this._handleClear();
+    }
+  }
+
+  // Handle operand
+  _handleOperand(value) {
+    if(this._activeOperand === '0') {
+      if(value === '0') {
+        return
+      } else if(value !== '.') {
+        this._activeOperand = '';
+      }
+    }
+
+    if(this._activeOperand.length < Calc.MAX_NUMBER) {
+      this._result = this._activeOperand += value;
+    }
+  }
+
+  // Handle operator
+  _handleOperator(operator) {
+    this._copyOperands();
+
+    switch(operator) {
+      case '1/x':
+      case 'sqrt':
+      case 'pow2':
+      case 'pow3':
+      case '10pow':
+      case 'factorial':
+        singleMath.apply(this);
+        break;
+      case '+':
+      case '-':
+      case '*':
+      case '/':
+      case 'powy':
+        simpleMath.apply(this);
+        break;
+      case '%':
+        complexMath.apply(this);
+        break;
+    }
+
+    // Simple math function
+    function simpleMath() {
```

Comment:
if you use this function with calculator as a context you'd better move it to _method in class declaration